[CI] Build failures with ABSEIL 20240116 and CMAKE 3.30

CI started to fail for this build:

```
CI / CMake test (with abseil) (pull_request) 
```

The failure is in:

```
./ci/install_abseil.sh
```

Failure:

```
-- Configuring done (0.5s)
CMake Error at CMake/AbseilHelpers.cmake:317 (target_link_libraries):
  The link interface of target "test_allocator" contains:

    GTest::gmock

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.

Call Stack (most recent call first):
  absl/container/CMakeLists.txt:206 (absl_cc_library)


CMake Generate step failed.  Build files cannot be regenerated correctly.
-- Generating done (0.2s)
Error: Process completed with exit code 1.
```

To investigate, I suspect a regression caused by the new image `Version: 20240708.1.0`:

```
Current runner version: '2.317.0'
Operating System
  Ubuntu
  20.04.6
  LTS
Runner Image
  Image: ubuntu-20.04
  Version: 20240708.1.0
```

Previous runs that builds properly:

```
Current runner version: '2.317.0'
Operating System
  Ubuntu
  20.04.6
  LTS
Runner Image
  Image: ubuntu-20.04
  Version: 20240701.1.0
```

See related:

* https://github.com/actions/runner-images/pull/10191

To investigate, CMake was upgraded.

--- a/ci/fix-abseil-cpp-issue-1536.patch
+++ b/ci/fix-abseil-cpp-issue-1536.patch
@@ -1,0 +1,23 @@
+commit 779a3565ac6c5b69dd1ab9183e500a27633117d5
+Author: Derek Mauro <dmauro@google.com>
+Date:   Tue Jan 30 10:13:25 2024 -0800
+
+    Avoid export of testonly target absl::test_allocator in CMake builds
+    
+    Closes #1536
+    
+    PiperOrigin-RevId: 602764437
+    Change-Id: Ia5c20a3874262a2ddb8797f608af17d7e86dd6d6
+
+diff --git a/absl/container/CMakeLists.txt b/absl/container/CMakeLists.txt
+index 449a2cad..ee9ca9c3 100644
+--- a/absl/container/CMakeLists.txt
++++ b/absl/container/CMakeLists.txt
+@@ -213,6 +213,7 @@ absl_cc_library(
+   DEPS
+     absl::config
+     GTest::gmock
++  TESTONLY
+ )
+ 
+ absl_cc_test(

--- a/ci/install_abseil.sh
+++ b/ci/install_abseil.sh
@@ -49,9 +49,10 @@ fi
 #
 # TODO(marcalff) Cleanup once abseil is upgraded to the next LTS
 
-echo "Patching abseil"
 
-patch -p1 << EOF
+if [ "${ABSEIL_CPP_VERSION}" -eq "20240116.1" ] || [ "${ABSEIL_CPP_VERSION}" -eq "20240116.2" ]; then
+  echo "Patching abseil"
+  patch -p1 << EOF
 commit 779a3565ac6c5b69dd1ab9183e500a27633117d5
 Author: Derek Mauro <dmauro@google.com>
 Date:   Tue Jan 30 10:13:25 2024 -0800
@@ -76,6 +77,9 @@ index 449a2cad..ee9ca9c3 100644
  
  absl_cc_test(
 EOF
+else
+  echo "Not patching abseil"
+fi
 
 mkdir build && pushd build
 cmake "${ABSEIL_CPP_BUILD_OPTIONS[@]}" ..

--- a/ci/install_abseil.sh
+++ b/ci/install_abseil.sh
@@ -55,7 +55,6 @@ fi
 if [ "${ABSEIL_CPP_VERSION}" = "20240116.1" ] || [ "${ABSEIL_CPP_VERSION}" = "20240116.2" ]; then
   echo "Patching abseil"
   patch -p1 < ${TOPDIR}/ci/fix-abseil-cpp-issue-1536.patch
-EOF
 else
   echo "Not patching abseil"
 fi

--- a/ci/install_abseil.sh
+++ b/ci/install_abseil.sh
@@ -50,7 +50,7 @@ fi
 # TODO(marcalff) Cleanup once abseil is upgraded to the next LTS
 
 
-if [ "${ABSEIL_CPP_VERSION}" -eq "20240116.1" ] || [ "${ABSEIL_CPP_VERSION}" -eq "20240116.2" ]; then
+if [ "${ABSEIL_CPP_VERSION}" = "20240116.1" ] || [ "${ABSEIL_CPP_VERSION}" = "20240116.2" ]; then
   echo "Patching abseil"
   patch -p1 << EOF
 commit 779a3565ac6c5b69dd1ab9183e500a27633117d5

--- a/ci/install_abseil.sh
+++ b/ci/install_abseil.sh
@@ -7,6 +7,8 @@ set -ex
 export DEBIAN_FRONTEND=noninteractive
 [ -z "${ABSEIL_CPP_VERSION}" ] && export ABSEIL_CPP_VERSION="20240116.1"
 
+TOPDIR=`pwd`
+
 BUILD_DIR=/tmp/
 INSTALL_DIR=/usr/local/
 pushd $BUILD_DIR
@@ -52,30 +54,7 @@ fi
 
 if [ "${ABSEIL_CPP_VERSION}" = "20240116.1" ] || [ "${ABSEIL_CPP_VERSION}" = "20240116.2" ]; then
   echo "Patching abseil"
-  patch -p1 << EOF
-commit 779a3565ac6c5b69dd1ab9183e500a27633117d5
-Author: Derek Mauro <dmauro@google.com>
-Date:   Tue Jan 30 10:13:25 2024 -0800
-
-    Avoid export of testonly target absl::test_allocator in CMake builds
-    
-    Closes #1536
-    
-    PiperOrigin-RevId: 602764437
-    Change-Id: Ia5c20a3874262a2ddb8797f608af17d7e86dd6d6
-
-diff --git a/absl/container/CMakeLists.txt b/absl/container/CMakeLists.txt
-index 449a2cad..ee9ca9c3 100644
---- a/absl/container/CMakeLists.txt
-+++ b/absl/container/CMakeLists.txt
-@@ -213,6 +213,7 @@ absl_cc_library(
-   DEPS
-     absl::config
-     GTest::gmock
-+  TESTONLY
- )
- 
- absl_cc_test(
+  patch -p1 < ${TOPDIR}/ci/fix-abseil-cpp-issue-1536.patch
 EOF
 else
   echo "Not patching abseil"

--- a/ci/install_abseil.sh
+++ b/ci/install_abseil.sh
@@ -22,6 +22,61 @@ if [ ! -z "${CXX_STANDARD}" ]; then
     ABSEIL_CPP_BUILD_OPTIONS=(${ABSEIL_CPP_BUILD_OPTIONS[@]} "-DCMAKE_CXX_STANDARD=${CXX_STANDARD}")
 fi
 
+#
+# ABSEIL_CPP_VERSION="20240116.1" fails to build with CMake 3.30
+# ABSEIL_CPP_VERSION="20240116.2" fails to build with CMake 3.30
+# note that somehow the same builds with CMake 3.29.6
+#
+# Error reported:
+#   CMake Error at CMake/AbseilHelpers.cmake:317 (target_link_libraries):
+#    The link interface of target "test_allocator" contains:
+#
+#      GTest::gmock
+#
+#    but the target was not found.  Possible reasons include:
+#
+#      * There is a typo in the target name.
+#      * A find_package call is missing for an IMPORTED target.
+#      * An ALIAS target is missing.
+#
+#   Call Stack (most recent call first):
+#    absl/container/CMakeLists.txt:206 (absl_cc_library)
+#
+# Root cause:
+#   https://github.com/abseil/abseil-cpp/pull/1536
+#
+# Applying fix from abseil commit 779a3565ac6c5b69dd1ab9183e500a27633117d5
+#
+# TODO(marcalff) Cleanup once abseil is upgraded to the next LTS
+
+echo "Patching abseil"
+
+patch -p1 << EOF
+commit 779a3565ac6c5b69dd1ab9183e500a27633117d5
+Author: Derek Mauro <dmauro@google.com>
+Date:   Tue Jan 30 10:13:25 2024 -0800
+
+    Avoid export of testonly target absl::test_allocator in CMake builds
+    
+    Closes #1536
+    
+    PiperOrigin-RevId: 602764437
+    Change-Id: Ia5c20a3874262a2ddb8797f608af17d7e86dd6d6
+
+diff --git a/absl/container/CMakeLists.txt b/absl/container/CMakeLists.txt
+index 449a2cad..ee9ca9c3 100644
+--- a/absl/container/CMakeLists.txt
++++ b/absl/container/CMakeLists.txt
+@@ -213,6 +213,7 @@ absl_cc_library(
+   DEPS
+     absl::config
+     GTest::gmock
++  TESTONLY
+ )
+ 
+ absl_cc_test(
+EOF
+
 mkdir build && pushd build
 cmake "${ABSEIL_CPP_BUILD_OPTIONS[@]}" ..
 make -j $(nproc)

--- a/tools/format.sh
+++ b/tools/format.sh
@@ -23,8 +23,8 @@ fi
 # No CRLF line endings, except Windows files.
 "${SED[@]}" 's/\r$//' $($FIND -name '*.ps1' -prune -o \
   -name '*.cmd' -prune -o -type f -print)
-# No trailing spaces.
-"${SED[@]}" 's/ \+$//' $($FIND -type f -print)
+# No trailing spaces, except in patch.
+"${SED[@]}" 's/ \+$//' $($FIND -name "*.patch" -prune -o -type f -print)
 
 # If not overridden, try to use clang-format-18 or clang-format.
 if [[ -z "$CLANG_FORMAT" ]]; then

